include/swift/AST/DiagnosticConsumer.h

```diff
@@ -136,32 +136,110 @@ class NullDiagnosticConsumer : public DiagnosticConsumer {
+    /// be associated with. An empty string means that a consumer is not
+    /// associated with any particular buffer, and should only receive
+    /// diagnostics that are not in any of the other consumers' files.
+    std::string inputFileName;
+
+    /// The consumer (if any) for diagnostics associated with the inputFileName.
+    /// A null pointer for the DiagnosticConsumer means that diagnostics for
+    /// this file should not be emitted.
+    std::unique_ptr<DiagnosticConsumer> consumer;
+
+    // Has this subconsumer ever handled a diagnostic that is an error?
+    bool hasAnErrorBeenConsumed = false;
+
+  public:
+    std::string getInputFileName() const { return inputFileName; }
+
+    DiagnosticConsumer *getConsumer() const { return consumer.get(); }
+
+    Subconsumer(std::string inputFileName,
+                std::unique_ptr<DiagnosticConsumer> consumer)
+        : inputFileName(inputFileName), consumer(std::move(consumer)) {}
+
+    void handleDiagnostic(SourceManager &SM, SourceLoc Loc,
+                          DiagnosticKind Kind,
+                          StringRef FormatString,
+                          ArrayRef<DiagnosticArgument> FormatArgs,
+                          const DiagnosticInfo &Info) {
+      if (!getConsumer())
+        return; // Suppress non-primary diagnostic in batch mode.
+      hasAnErrorBeenConsumed |= Kind == DiagnosticKind::Error;
+      getConsumer()->handleDiagnostic(SM, Loc, Kind, FormatString, FormatArgs,
+                                      Info);
+    }
+    
+    void informDriverOfIncompleteBatchModeCompilation() {
+      if (!hasAnErrorBeenConsumed && getConsumer())
+        getConsumer()->informDriverOfIncompleteBatchModeCompilation();
+    }
+  };
 
 private:
   /// All consumers owned by this FileSpecificDiagnosticConsumer.
-  const SmallVector<ConsumerPair, 4> SubConsumers;
+  SmallVector<Subconsumer, 4> Subconsumers;
 
 public:
   // The commented-out consts are there because the data does not change
   // but the swap method gets called on this structure.
-  struct ConsumerSpecificInformation {
+  class ConsumerSpecificInformation {
+  private:
+    /// The range of SourceLoc's for which diagnostics should be directed to
+    /// this subconsumer.
     /*const*/ CharSourceRange range;
-    /// The DiagnosticConsumer may be empty if those diagnostics are not to be
-    /// emitted.
-    DiagnosticConsumer * /*const*/ consumer;
-    bool hasAnErrorBeenEmitted = false;
 
+    /// Index into Subconsumers vector for this subconsumer.
+    /*const*/ unsigned subconsumerIndex;
+
+  public:
     ConsumerSpecificInformation(const CharSourceRange range,
-                                DiagnosticConsumer *const consumer)
-        : range(range), consumer(consumer) {}
+                                unsigned subconsumerIndex)
+        : range(range), subconsumerIndex(subconsumerIndex) {}
+
+    Subconsumer &subconsumer(FileSpecificDiagnosticConsumer &c) const {
```

Comment:
I see why you did this: it lets you keep the `subconsumerIndex` field private, and makes the code sort of look like there's a `subconsumer` computed property. But having to pass in the owner breaks the illusion. I think I'd prefer something like `operator[](const ConsumerSpecificInformation &)` on FileSpecificDiagnosticConsumer if you want to keep `subconsumerIndex` out of client code.

Reply:
OK. Requires an accessor for `subconsumerIndex` to guard against mutation in public, but it does read better.